File: BOJ/2636_BFS.cpp
#include <iostream>
#include <vector>
#include <queue>

// boj 2636 치즈, 골드 5, BFS 사용 
using namespace std;

int dy[4] = {-1,0,1,0};
int dx[4] = {0,1,0,-1};

vector<vector<int>> plate;

pair<int,int> countTime(int R, int C, int cheese){
    int time = 0;
    int lastCheese = 0;

    while (cheese>0){
        vector<vector<bool>> visited(R, vector<bool>(C, false));
        queue<pair<int, int>> blank;

        visited[0][0] = true;
        blank.push({0,0});

        vector<pair<int, int>> edge;

        while (!blank.empty()){
            pair<int, int> cur = blank.front();
            blank.pop();

            for (int i = 0; i < 4; ++i) {
                int nr = cur.first + dy[i];
                int nc = cur.second + dx[i];

                if (nr>=0 && nr<R && nc>=0 && nc<C && !visited[nr][nc]){
                    if (plate[nr][nc]==0){
                        visited[nr][nc] = true;
                        blank.push({nr, nc});
                    }else{
                        visited[nr][nc] = true;
                        edge.push_back({nr, nc});
                    }
                }
            }
        }

        for (int i = 0; i < edge.size(); ++i) {
            plate[edge[i].first][edge[i].second] = 0;
        }
        if (edge.size() >= cheese) lastCheese = cheese;
        cheese -= edge.size();
        time++;
    }

    return make_pair(time, lastCheese);
}

int main() {
    int R, C;
    cin>>R>>C;

    int cheese = 0;

    plate.assign(R, vector<int>(C, -1));
    for (int i = 0; i < R; ++i) {
        for (int j = 0; j < C; ++j) {
            cin>>plate[i][j];
            if (plate[i][j] == 1) cheese++;
        }
    }

    pair<int, int> answer = countTime(R, C, cheese);

    cout<<answer.first<<'\n';
    cout<<answer.second<<'\n';

    return 0;
}

File: BOJ/2636_BFS.swift
import Foundation
// boj 2636 치즈, 골드 4, bfs

struct Queue<T> {
    private var queue: [T] = []
    
    public var count: Int {
        return queue.count
    }
    
    public var front : T {
        return queue[0]
    }
    
    public var isEmpty: Bool {
        return queue.isEmpty
    }
    
    public mutating func enqueue(_ element: T) {
        queue.append(element)
    }
    
    public mutating func dequeue() -> T? {
        return isEmpty ? nil : queue.removeFirst()
    }
}

let st1 = readLine()!.split(separator: " ").map{Int($0)!}
let N = st1[0]
let M = st1[1]

let dy = [-1,0,1,0]
let dx = [0,1,0,-1]

var map = [[Int]](repeating: [Int](repeating: 0, count: M), count: N)
var cheese = 0
var time = 0
var deleted = 0

for i in 0..<N {
    let st2 = readLine()!.split(separator: " ").map{Int($0)!}
    for j in 0..<M {
        map[i][j] = st2[j]
        if map[i][j] == 1 { cheese += 1 }
    }
}

func bfs(_ startR: Int, _ startC: Int, _ visited: inout [[Bool]], _ newMap: inout [[Int]]) {
    var q = Queue<(Int, Int)>()
    
    q.enqueue((startR, startC))
    visited[startR][startC] = true
    
    while !q.isEmpty {
        guard let cur = q.dequeue() else { continue }
        
        for i in 0..<4 {
            let ny = cur.0 + dy[i]
            let nx = cur.1 + dx[i]
            
            if ny<0 || ny>=N || nx<0 || nx>=M || visited[ny][nx] { continue }
            
            visited[ny][nx] = true
            
            if map[ny][nx] == 0 {
                q.enqueue((ny, nx))
            } else {
                newMap[ny][nx] = 0
                deleted += 1
            }
        }
    }
    
}

while cheese > 0 {
    
    time += 1
    
    var newMap = map
    var visited = [[Bool]](repeating: [Bool](repeating: false, count: M), count: N)
    
    deleted = 0
    bfs(0, 0, &visited, &newMap)
    
    cheese -= deleted
    map = newMap
}

print("\(time)\n\(deleted)")
